### PR TITLE
Use features instead of properties in `bbox_annotator` example

### LINKED ISSUE #4079: Add features Event to layers
Add an `Event` named `features` to all layers that currently have a `features` property. Then emit that event when the value that the getter would return has changed.

### LINKED ISSUE #4166: Update examples/bbox_annotator.py to use modern features architecture
## 🧰 Task

[`examples/bbox_annotator.py`](https://github.com/napari/napari/blob/c76b1f8946259c8e728a153b130279238eacf1ec/examples/bbox_annotator.py) uses the deprecated `default_properties` and `current_properties` APIs. We need to decide how to move forward using pandas categorical features. See this comment by @andy-sweet:

https://github.com/napari/napari/pull/4162#issuecomment-1054881453

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
This mostly updates the `bbox_annotator` example to use features instead of properties.

Setting up a `features` dataframe with a categorical dtype is a simple and hopefully uncontentious way of replacing `default_properties`.

Replacing `current_properties` with `feature_defaults` is slightly trickier. After some discussion, I decided the simplest way forward here is to add `Layer.events.feature_defaults` and require the user to emit that event when they modify `Layer.feature_defaults` in-place. The alternative approach is to define a `feature_defaults` setter, but that has a few downsides.

- We need to define the logic for the setter. E.g. what happens if the columns of the new feature defaults are different to those of the existing features?
- Requires a little more code.
- We need to decide if we want to keep the selection-dependent behavior of the `current_properties` setter.  

We could add a feature defaults setter in the future, but we would need to resolve these questions.

## Type of change
- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Closes #4166
Closes #4079  

# How has this been tested?
- [ ] existing tests pass with my change
- [ ] manual use of the `bbox_annotator.py` example seems to work as intended

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: examples/bbox_annotator.py
Comment (andy-sweet) [inline examples/bbox_annotator.py]: @jni : I'm curious what you think of this approach to mutating feature defaults.

At first I added the `feature_defaults` setter (as in this PR) and used it successfully, but then I thought about this alternative (mutating in-place and emitting the event) and kinda like it. What do you think?

Comment (sofroniewn) [reply examples/bbox_annotator.py]: This was discussed in team meeting a bit - I can see pros and cons. I'm ok with it, it has a comment - though I might add a more explicit comment about the need to emit the event as it is the only place currently required. The main reason I like this approach is that if we ever get to a world where we don't need to manually emit the event it is very easy to migrate people (just get them to drop extra line). I don't have strong opinions though, so if @jni or @tlambert03 have strong views I will defer

**Review by sofroniewn — APPROVED** [review]
LGTM! Though agreed - probably good to have others weigh in about that event emission line above

File: examples/bbox_annotator.py
Comment (tlambert03) [reply examples/bbox_annotator.py]: I'm fine with it.  No strong opinions either.  wish it weren't necessary, but it is! :)

**Review by tlambert03 — APPROVED** [review]

File: examples/bbox_annotator.py
Comment (andy-sweet) [inline examples/bbox_annotator.py:46]: I redefined this callback function to not take an `event` argument to avoid the following warning.

```
examples/bbox_annotator.py:69: FutureWarning: 

magicgui 0.4.0 will change the way that callbacks are called.
Instead of a single `Event` instance, with an `event.value` attribute,
callbacks will receive the value(s) directly:

@widget.changed.connect
def my_callback(*args):
    # *args are the value(s) themselves!

To silence this warning you may either provide a callback that has more
or less than 1 parameter.  Or annotate the single parameter as anything
*other* than `Event`, e.g. `def callback(x: int): ...`
For details, see: https://github.com/napari/magicgui/issues/255
  label_menu.changed.connect(label_changed)
```

Instead I just look up the new value as `label_menu.value`. Let me know if I should do something different.

**Comment by andy-sweet** [comment]
As discussed, I updated this PR so that it does not define a `feature_defaults` setter, but instead adds events for `features` and `feature_defaults` which can be emitted and connected to instead.

Probably worth at least one person taking another look before merging.

File: examples/bbox_annotator.py
Comment (jni) [reply examples/bbox_annotator.py:46]: I'd say let's have the events carry the information. I prefer callbacks to be self-contained rather than rely on state. So you could silence the warning with just:

```python
def label_changed(value: str):  # yes, annoyingly, pandas doesn't return Categorical, but str, from the series
    ...
```

My *actual* preference would be to use `toolz.curry` to avoid having to make closures using nested function definitions. So:

```python
import toolz as tz

@tz.curry
def set_default_feature(layer, feature_name, value):
    layer.feature_defaults[feature_name] = value
    layer.events.feature_defaults(feature_defaults)
    # optionally we could add feature_names here, containing
    # a list of which feature names have changed defaults

...

label_menu.changed.connect(set_default_feature(shapes_layer, label_property))
# and I would rename `label_property` to `label_feature_name` maybe?
```

What do you think? Currying is definitely advanced, but so is this example so 🤷. Either way I prefer less state even if we can't get to no state.

**Review by jni — CHANGES_REQUESTED** [review]
@andy-sweet thanks for the ping about this during meeting. I've left a suggestion for an alternative approach. In general we've found that we would like events to carry *more* information rather than less, so I'd prefer if we started doing that here.

Let me know what you think. I'm definitely not wedded to the curried approach as it is non-standard, but a more standard way would be to use functools.partial. Either way the idea of de-nesting functions and thus making them more general appeals to me.

File: examples/bbox_annotator.py
Comment (andy-sweet) [reply examples/bbox_annotator.py:46]: Agreed that using `def label_changed(value: str)` is an improvement because it relies on less captured state in the callback, so I pushed that.

I'm less sure about using partials in this example. On the plus side, I'm always a fan of less indentation :) But using them in this example doesn't actually decrease the state we're capturing. Either that state is being captured when defining the nested function or when doing the partial evaluation. If we wanted to reuse those functions (e.g. `set_default_feature`) elsewhere with capturing state, then I see the improvement here. But this example doesn't do that. Maybe you could argue these functions could be copied elsewhere by plugin devs and be more reusable then, but I'm unsure how true that will be in practice.

Anyway, I tried it out and got it working with some wrinkles. I didn't push yet, so here's [a gist instead](https://gist.github.com/andy-sweet/d63c46359ecfd905877298d921b96ece). As you alluded to, we want to pass the new default values with the `feature_defaults` event. That could just be a reference to `Shapes.feature_defaults`, some subset of its columns, or maybe just a plain `dict`. But we have to pick if we're using partials in this example, and we have to pass those values whenever those events are emitted (e.g. in the `Shapes` layer).

As we are adding the `feature_defaults` and `features` events in this PR, now is certainly a good time to think about what values we pass with the event. But I think it's also OK to pass nothing for now and add values later (that won't break people right?).

Comment (jni) [reply examples/bbox_annotator.py:46]: > As we are adding the feature_defaults and features events in this PR, now is certainly a good time to think about what values we pass with the event. But I think it's also OK to pass nothing for now and add values later (that won't break people right?).

Well, it will, if their callback takes no arguments and we start pass arguments, right? If their callback takes `*args` that's true, but I don't see a priori why they would do that.

Comment (jni) [reply examples/bbox_annotator.py:46]: I honestly really like the gist! But I'm wondering why it takes an Event rather than the event value directly? Which I thought is what started this whole discussion? We can bump the magicgui dependency if needed...

Comment (jni) [reply examples/bbox_annotator.py:46]: @tlambert03 do you have thoughts about the two patterns here (nesting vs partials)?

Comment (andy-sweet) [reply examples/bbox_annotator.py:46]: > Well, it will, if their callback takes no arguments and we start pass arguments, right?

Yep, I think you're right. I was probably thinking about [Qt signals and slots](https://doc.qt.io/qt-5/signalsandslots.html), where I think the slot can have fewer arguments than the signal.

Comment (andy-sweet) [reply examples/bbox_annotator.py:46]: > But I'm wondering why it takes an Event rather than the event value directly?

Good catch. Here's the error when I try to pass new/changed feature defaults directly rather than the `Event`: https://gist.github.com/andy-sweet/12fc070b25e388232f2ff12a599c31d1

I don't immediately understand the traceback, so maybe you or @tlambert03 can take a look. Note that upgrading the type of the final argument from `Any` to `pd.DataFrame` does not help. My environment has `magicgui==0.3.6`, which is the min required in napari's `setup.cfg`.

Also note that passing the `Event` does not cause the aforementioned warnings. Again, I don't entirely understand why, as the partial does only have one argument, so it seems like the issue I first encountered. Maybe some introspection in the event code is interacting in an interesting way with the partial?

Comment (andy-sweet) [reply examples/bbox_annotator.py:46]: > > Well, it will, if their callback takes no arguments and we start pass arguments, right?
> 
> Yep, I think you're right. I was probably thinking about [Qt signals and slots](https://doc.qt.io/qt-5/signalsandslots.html), where I think the slot can have fewer arguments than the signal.

Actually, nope. Our events do seem to behave similarly to Qt, so the callback can have fewer arguments than the event emission. I.e. the following causes `on_features_changed` to be executed without an error.

```python
from napari.utils.events import EmitterGroup, Event

def on_features_changed():
    print('features changed')

group = EmitterGroup(features=Event)
group.features.connect(on_features_changed)
group.features(x=10)
```

I don't know if that's intended and may break soon. I guess we need @tlambert03 yet again.

**Comment by andy-sweet** [comment]
If we do want to add arguments to the new `features` and `feature_defaults` events, then that probably requires some significant design work so that those arguments are consistent with other arguments we add to events in the future. Some important questions that come to mind are the following.

1. Should the arguments be positional or keyword?
2. Should we define specific event types like `FeaturesChangedEvent` to help document what the arguments are?
3. For these features-related events, should we pass a reference to the whole new instance (e.g. `self.events.features(new_features=self.features)`), only the columns/rows that changed (e.g. `self.events.features(changed=self.features[['class', 'conf']][[1, 7, 10]])`? Maybe both?

It feels like these questions are growing out of scope of this PR. And I don't have much bandwidth to take the lead on them in the short term. If we can confirm that we can add arguments to events later without breaking people, I strongly suggest merging this and improving things later.

File: examples/bbox_annotator.py
Comment (andy-sweet) [reply examples/bbox_annotator.py:46]: > I don't immediately understand the traceback, so maybe you or @tlambert03 can take a look.

I pushed the alternative approach in the gist to a branch. See this draft PR on my fork if you want to try to change that to not use `Event` arguments in the callbacks: https://github.com/andy-sweet/napari/pull/4

Comment (tlambert03) [reply examples/bbox_annotator.py:46]: just catching up:
If this is a callback for a magicgui change event (and not an emitter from napari), then, to confirm some of the above: 
1. it's a psygnal based callback (not one from vispy/napari's events.py)
2. like Qt, it _can_ [accept fewer arguments](https://github.com/tlambert03/psygnal#extra-positional-arguments-ignored) than the emitter is providing.  That _is_ intended and well-tested.

... what was the key question about partials here?   the following seems good to me if it meets your needs
```python
@ label_menu.changed.connect
def label_changed(value: str): ...
```

Comment (tlambert03) [reply examples/bbox_annotator.py:46]: sorry, and also, the napari `event.py` also knows whether or not to pass the event object to the callback too (see the `_callback_pass_event` store in event.py)

Comment (jni) [reply examples/bbox_annotator.py:46]: > sorry, and also, the napari event.py also knows whether or not to pass the event object to the callback too (see the _callback_pass_event store in event.py)

Fascinating! Has that always been the case? I remember early on futzing around with lambdas to make sure that the event emission and the callback signatures would match, with `lambda ev: func()`

> ... what was the key question about partials here?

```python
@label_menu.changed.connect
def label_changed(value: str): ...
```

This definition is nested within `label_menu` so that it has access to the viewer and layers, which are then used in a closure. My proposal is to un-nest it and pass those as arguments, then use partials (either with toolz.curry or functools.partial) to connect.

```python
def label_changed(viewer, layer, value: str):
    ...

...

label_menu.changed.connect(partial(label_changed, viewer, layer))
```

@andy-sweet 

> My environment has magicgui==0.3.6, which is the min required in napari's setup.cfg

We can potentially bump that dependency. I think this might need a newer magicgui.

**Review by jni — APPROVED** [review]
> It feels like these questions are growing out of scope of this PR. And I don't have much bandwidth to take the lead on them in the short term. If we can confirm that we can add arguments to events later without breaking people, I strongly suggest merging this and improving things later.

You're right, and indeed these are long-running questions, see e.g. #2809. I agree that we can drop that aspect of this PR. Still leaves the nesting, which we can also address later tbh. As mentioned I still prefer the approach using partials, but I won't block the merge for it. I'd just like @tlambert03 to comment on it ([here](https://github.com/napari/napari/pull/4218#discussion_r830694078)!), then pull the trigger on merge if he thinks I'm being annoyingly pedantic. :joy:

File: examples/bbox_annotator.py
Comment (andy-sweet) [reply examples/bbox_annotator.py:46]: > If this is a callback for a magicgui change event (and not an emitter from napari),

I hate extending this thread, but as in my example above (and in this PR), we also get the described behavior for napari events/emitters. Is that intended?

Comment (tlambert03) [reply examples/bbox_annotator.py:46]: yes, see https://github.com/napari/napari/pull/4218#discussion_r830659211

Comment (tlambert03) [reply examples/bbox_annotator.py:46]: > Fascinating! Has that always been the case? I remember early on futzing around with lambdas to make sure that the event emission and the callback signatures would match, with lambda ev: func()

no not always, @Czaki added that nice feature in https://github.com/napari/napari/pull/3449

**Comment by sofroniewn** [comment]
I'm just catching up here.

> I'd just like @tlambert03 to comment on it (https://github.com/napari/napari/pull/4218#discussion_r830694078!), then pull the trigger on merge if he thinks I'm being annoyingly pedantic

Seems like reply was that things have been this way for a bit. Maybe we should change some of this stuff at some point - i also feel like I agree with @jni that having explicit args if nicer, but that seems like a more significant change. I propose that we just merge this now as is and then @jni maybe you want to summarize some of the discussion about the args and current state in a new issue. If no one objects I'll merge in a day or two - thanks!!

File: examples/bbox_annotator.py
Comment (tlambert03) [reply examples/bbox_annotator.py:46]: > This definition is nested within label_menu so that it has access to the viewer and layers, which are then used in a closure. My proposal is to un-nest it and pass those as arguments, then use partials (either with toolz.curry or functools.partial) to connect.

I don't have a strong preference here either.  I think the main concern here is holding a strong reference, but in this case, I suspect _both_ the closure and the partial would hold a strong reference.  It's just an example though, so let's not overthink it :joy: they both work.

**Comment by tlambert03** [comment]
agreed!  let's just merge now.  I think everyone has weighed in

**Pull request merged**
